File: job-scraper-backend/scrapers/opportunity_scraper.py
from .base_scraper import BaseScraper
import requests
from .keywords import DEFAULT_KEYWORDS
import re
from datetime import datetime, timedelta

class OpportunityScraper(BaseScraper):
    """A scraper for 'opportunity.ini.rw' using their internal API."""

    def _is_deadline_valid(self, deadline_date_str):
        """
        Check if the job deadline is either future or within 2 weeks from today.
        Returns True if the deadline is valid, False otherwise.
        """
        if deadline_date_str == "N/A" or not deadline_date_str.strip():
            return True  # Include jobs without deadline information
        
        try:
            # Try to parse different date formats commonly used
            date_formats = [
                "%A, %B %d %Y",  # e.g., "Wednesday, May 28 2025"
                "%A, %b %d %Y",  # e.g., "Wednesday, May 28 2025" (short month)
                "%d %B %Y",      # e.g., "15 July 2025"
                "%d %b %Y",      # e.g., "15 Jul 2025"
                "%d-%m-%Y",      # e.g., "15-07-2025"
                "%d/%m/%Y",      # e.g., "15/07/2025"
                "%Y-%m-%d",      # e.g., "2025-07-15"
                "%B %d, %Y",     # e.g., "July 15, 2025"
                "%b %d, %Y",     # e.g., "Jul 15, 2025"
            ]
            
            deadline_date = None
            for date_format in date_formats:
                try:
                    deadline_date = datetime.strptime(deadline_date_str.strip(), date_format)
                    break
                except ValueError:
                    continue
            
            if deadline_date is None:
                # Try simple ISO format containing time as well if needed, though scraper usually gets date string
                try:
                    deadline_date = datetime.fromisoformat(deadline_date_str.replace("Z", "+00:00"))
                except ValueError:
                    pass

            if deadline_date is None:
                print(f"Warning: Could not parse date format: {deadline_date_str}")
                return True  # Include jobs with unparseable dates
            
            # Remove timezone info for comparison if present (naive vs aware)
            if deadline_date.tzinfo is not None:
                deadline_date = deadline_date.replace(tzinfo=None)

            # Calculate the threshold date (2 weeks ago from today)
            today = datetime.now()
            two_weeks_ago = today - timedelta(days=14)
            
            # Include jobs whose deadline is either in the future or within the last 2 weeks
            return deadline_date >= two_weeks_ago
            
        except Exception as e:
            print(f"Error parsing deadline date '{deadline_date_str}': {e}")
            return True  # Include jobs with date parsing errors

    def scrape(self, keyword=None):
        # The specific API endpoint found
        base_url = "https://opportunityapi.ini.rw/api/opportunities"
        
        # Parameters based on the request inspection
        params = {
            "locale": "en",
            "limit": 200,  # Request 200 items to minimize pagination needs
            "offset": 0,
            "status": "approved",
            "sort": "default_ranking",
            # If a specific keyword is requested, pass it here, otherwise empty
            "q": keyword if keyword else "" 
        }

        # Behave like a browser
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
            "Accept": "application/json",
            "Origin": "https://opportunity.ini.rw",
            "Referer": "https://opportunity.ini.rw/"
        }

        print(f"Fetching data from {base_url}...")

        try:
            response = requests.get(base_url, params=params, headers=headers, timeout=10)
            
            if response.status_code != 200:
                print(f"Failed to fetch data: Status {response.status_code}")
                return {"total_jobs": 0, "unique_companies": 0, "jobs": []}

            data = response.json()
            
            # The API returns a list directly based on the observed behavior, 
            # or a dict. We'll handle both just in case.
            raw_jobs = []
            if isinstance(data, list):
                raw_jobs = data
            elif isinstance(data, dict):
                # Common keys for paginated APIs
                if 'results' in data:
                    raw_jobs = data['results']
                elif 'data' in data:
                    raw_jobs = data['data']
                elif 'items' in data:
                    raw_jobs = data['items']
                else:
                    # Fallback: if dict has no obvious list, it might be the dict itself if malformed,
                    # but usually it's one of the above. 
                    # For opportunity.ini.rw, previous observation suggests it might be a list or standard paginated structure.
                    # We will assume 'results' or direct list based on standard REST practices if not specified.
                    # However, let's look at the structure if we can. 
                    # Since we can't see the response, we will robustly default to 'results' or check if data itself is iterable.
                    pass
            
            all_jobs = []
            company_names = set()

            for job in raw_jobs:
                # Handle nested company objects
                company_name = "Unknown"
                if isinstance(job.get('company'), dict):
                    company_name = job['company'].get('name', 'Unknown')
                elif isinstance(job.get('company'), str):
                    company_name = job['company']
                
                # Construct the full link using the slug
                # Pattern: https://opportunity.ini.rw/en/opportunities/{slug}
                slug = job.get('slug', '')
                if not slug:
                    continue # Skip if no slug

                full_link = f"https://opportunity.ini.rw/en/opportunities/{slug}"
                
                # Correct field mapping based on API inspection
                title = job.get('title_en', job.get('title_rw', 'No Title'))
                
                # Dates
                published_date = job.get('created_at', 'N/A')
                deadline_date = job.get('closing_date', 'N/A')
                
                # Location
                location = job.get('location_en', 'Rwanda')
                
                # Company (prefer direct name field if available)
                if job.get('company_name'):
                    company_name = job.get('company_name')
                elif isinstance(job.get('company'), dict):
                     company_name = job['company'].get('name', 'Unknown')
                elif isinstance(job.get('company'), str):
                    company_name = job['company']
                else:
                    company_name = "Unknown"

                all_jobs.append({
                    "title": title,
                    "company": company_name,
                    "link": full_link,
                    "published_date": published_date,
                    "deadline_date": deadline_date,
                    "location": location
                })
                
                if company_name != "Unknown":
                    company_names.add(company_name)

            # --- Filtering Logic (Client-side filtering for double safety) ---
            # Even though we passed 'q' to the API, 'keyword' in our app context usually implies 
            # strict filtering for "IT Jobs" when no specific keyword is passed.
            # If the user passed a keyword, the API 'q' param covered it.
            # If the user did NOT pass a keyword, we usually filter by DEFAULT_KEYWORDS (IT jobs) 
            # effectively turning the generic "all jobs" list into "IT jobs".
            
            if keyword:
                # API handled it, pass through (or refine if API fuzzy match is too broad)
                filtered_jobs = [
                    job for job in all_jobs 
                    if self._is_deadline_valid(job['deadline_date'])
                ]
            else:
                # Default behavior: Filter for IT keywords locally
                pattern = r'\b(' + '|'.join(DEFAULT_KEYWORDS) + r')\b'
                filtered_jobs = [
                    job for job in all_jobs 
                    if re.search(pattern, job['title'], re.IGNORECASE) and
                    self._is_deadline_valid(job['deadline_date'])
                ]

            return {
                "total_jobs": len(all_jobs), # Total fetched
                "unique_companies": len(company_names),
                "jobs": filtered_jobs # Return only relevant ones
            }

        except Exception as e:
            print(f"Error scraping Opportunity: {str(e)}")
            return {"total_jobs": 0, "unique_companies": 0, "jobs": []}

File: job-scraper-backend/src/scrapers/opportunityScraper.js
// src/scrapers/opportunityScraper.js

import axios from 'axios';
import { BaseScraper } from './baseScraper.js';
import { DEFAULT_KEYWORDS } from './keywords.js';

export class OpportunityScraper extends BaseScraper {
  /**
   * Scraper for 'opportunity.ini.rw' using their internal API.
   */

  _isDeadlineValid(deadlineDateStr) {
    if (deadlineDateStr === 'N/A' || !deadlineDateStr || !deadlineDateStr.trim()) {
      return true; // Include jobs without deadline information
    }

    try {
      let deadlineDate = null;
      const str = deadlineDateStr.trim();

      // Try Date.parse (handles ISO, standard formats)
      const parsed = Date.parse(str);
      if (!isNaN(parsed)) {
        deadlineDate = new Date(parsed);
      }

      if (!deadlineDate) {
        // Try dd-mm-yyyy
        const ddmmyyyy = str.match(/^(\d{1,2})-(\d{1,2})-(\d{4})$/);
        if (ddmmyyyy) {
          deadlineDate = new Date(
            parseInt(ddmmyyyy[3]),
            parseInt(ddmmyyyy[2]) - 1,
            parseInt(ddmmyyyy[1])
          );
        }
      }

      if (!deadlineDate) {
        // Try dd/mm/yyyy
        const ddmmyyyySlash = str.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
        if (ddmmyyyySlash) {
          deadlineDate = new Date(
            parseInt(ddmmyyyySlash[3]),
            parseInt(ddmmyyyySlash[2]) - 1,
            parseInt(ddmmyyyySlash[1])
          );
        }
      }

      if (!deadlineDate || isNaN(deadlineDate.getTime())) {
        console.log(`Warning: Could not parse date format: ${deadlineDateStr}`);
        return true; // Include jobs with unparseable dates
      }

      const today = new Date();
      const twoWeeksAgo = new Date(today.getTime() - 14 * 24 * 60 * 60 * 1000);

      return deadlineDate >= twoWeeksAgo;
    } catch (err) {
      console.error(`Error parsing deadline date '${deadlineDateStr}':`, err);
      return true;
    }
  }

  async fetchAll() {
    const baseUrl = 'https://opportunityapi.ini.rw/api/opportunities';

    const params = {
      locale: 'en',
      limit: 200,
      offset: 0,
      status: 'approved',
      sort: 'default_ranking',
      q: '',
    };

    const headers = {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
      Accept: 'application/json',
      Origin: 'https://opportunity.ini.rw',
      Referer: 'https://opportunity.ini.rw/',
    };

    console.log(`Fetching data from ${baseUrl}...`);

    let data;
    try {
      const response = await axios.get(baseUrl, {
        params,
        headers,
        timeout: 10000,
      });

      if (response.status !== 200) {
        console.log(`Failed to fetch data: Status ${response.status}`);
        return { allJobs: [], companyNames: [] };
      }

      data = response.data;
    } catch (err) {
      console.error(`Error fetching from API: ${err.message}`);
      return { allJobs: [], companyNames: [] };
    }

    // The API returns a list directly or a dict with results/data/items
    let rawJobs = [];
    if (Array.isArray(data)) {
      rawJobs = data;
    } else if (data && typeof data === 'object') {
      if (data.results) rawJobs = data.results;
      else if (data.data) rawJobs = data.data;
      else if (data.items) rawJobs = data.items;
    }

    const allJobs = [];
    const companyNames = new Set();

    for (const job of rawJobs) {
      const slug = job.slug;
      if (!slug) continue;

      const fullLink = `https://opportunity.ini.rw/en/opportunities/${slug}`;
      const title = job.title_en || job.title_rw || 'No Title';

      const publishedDate = job.created_at || 'N/A';
      const deadlineDate = job.closing_date || 'N/A';
      const location = job.location_en || 'Rwanda';

      let companyName = 'Unknown';
      if (job.company_name) {
        companyName = job.company_name;
      } else if (job.company && typeof job.company === 'object') {
        companyName = job.company.name || 'Unknown';
      } else if (typeof job.company === 'string') {
        companyName = job.company;
      }

      allJobs.push({
        title,
        company: companyName,
        link: fullLink,
        published_date: publishedDate,
        deadline_date: deadlineDate,
        location,
      });

      if (companyName !== 'Unknown') {
        companyNames.add(companyName);
      }
    }

    return { allJobs, companyNames: [...companyNames] };
  }

  /**
   * Override filterJobs to also check deadline validity.
   */
  filterJobs(allJobs, keyword = null, customKeywords = null) {
    const filterKeywords = customKeywords || DEFAULT_KEYWORDS;
    let filteredJobs;
    if (keyword) {
      // Single keyword: match title + deadline check
      const pattern = new RegExp(
        `\\b${keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`,
        'i'
      );
      filteredJobs = allJobs.filter(
        (job) => pattern.test(job.title) && this._isDeadlineValid(job.deadline_date)
      );
    } else {
      const pattern = new RegExp(`\\b(${filterKeywords.join('|')})\\b`, 'i');
      filteredJobs = allJobs.filter(
        (job) => pattern.test(job.title) && this._isDeadlineValid(job.deadline_date)
      );
    }
    return filteredJobs;
  }
}
